### PR TITLE
Implement formatting for fractional seconds

### LINKED ISSUE #485: Improve FractionalSecond
#445 introduced a placeholder for fractional seconds.  It should be improved by:

- [x] Enforce range limits on the values
- [X] Fully support in pattern formatting
- [x] Add to MockDateTime FromStr
- [X] Add to day period resolution logic (https://github.com/unicode-org/icu4x/pull/445#discussion_r570612228)

Things to keep in mind:

- Most formats for fractional seconds appear as `"s.SSS"`; however, the number of fractional seconds should be programmer-defined.
- See #565 for a previous attempt.

### PR DESCRIPTION
**Pull request by dminor**
Fixes #485

### CODE REVIEW

**Review by gregtatum — CHANGES_REQUESTED** [review]
I'm marking request changes for 3 reasons:

1. Potential arithmetic overflow for the API `FieldLength::Fixed(n)`
2. No max field size enforced from the parsed pattern, which I believe could lead to arithmetic overflow as well.
3. The `unimplemented` panics.

File: components/datetime/src/pattern/reference/parser.rs [hunk 1]
```diff
@@ -290,6 +290,22 @@ mod tests {
...
+                vec![
+                    (fields::Hour::H23.into(), FieldLength::TwoDigit).into(),
+                    ':'.into(),
+                    (FieldSymbol::Minute, FieldLength::TwoDigit).into(),
+                    ':'.into(),
+                    (fields::Second::Second.into(), FieldLength::TwoDigit).into(),
+                    '.'.into(),
+                    (
+                        fields::Second::FractionalSecond.into(),
+                        FieldLength::Fixed(2),
```
[the remark this answers is not in the document]
Comment (dminor) [reply components/datetime/src/pattern/reference/parser.rs:304]: Writing the test here was important for me to be able to debug and understand the parsing behaviour. I hate to remove a test once written, which is why I left it here.

File: components/datetime/tests/patterns/tests/dayperiods.json [hunk 2]
```diff
@@ -477,6 +477,36 @@
...
+          },
+          {
+            "patterns": [
+              "h:mm:ss.SSSSSSSS"
+            ],
+            "expected": "1:00:00.10101000"
+          },
+          {
+            "patterns": [
+              "h:mm:ss.SSSSSSSSSSS"
```
[the remark this answers is not in the document]
Comment (dminor) [reply components/datetime/tests/patterns/tests/dayperiods.json:505]: There's a place for testing errors in parser.rs. I've added a test there. Here I'll put a test for 127 `S`, the largest the implementation supports without generating an error.

File: components/datetime/src/pattern/reference/parser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gregtatum) [reply components/datetime/src/pattern/reference/parser.rs:304]: Sorry, I wasn't clear in what I was saying. I'm happy to keep this test here. 👍 I too find these types of tests useful for debugging.

**Review by gregtatum — APPROVED** [review]
Thanks for fixing the overflows and panics, this looks good to me now 👍

**Pull request merged**